src/category_theory/essential_image.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import category_theory.natural_isomorphism
+import category_theory.full_subcategory
+
+/-!
+# Essential image of a functor
+
+The essential image `ess_image` of a functor consists of the objects in the target category which
+are isomorphic to an object in the image of the object function.
+This, for instance, allows us to talk about objects belonging to a subcategory expressed as a
+functor rather than a subtype, preserving the principle of equivalence. For example this lets us
+define exponential ideals.
+
+The essential image can also be seen as a subcategory of the target category, and witnesses that
+a functor decomposes into a essentially surjective functor and a fully faithful functor.
+(TODO: show that this decomposition forms an orthogonal factorisation system).
+-/
+universes v₁ v₂ u₁ u₂
+
+noncomputable theory
+
+namespace category_theory
+
+variables {C : Type u₁} {D : Type u₂} [category.{v₁} C] [category.{v₂} D] {F : C ⥤ D}
+
+namespace functor
+
+/--
+The essential image of a functor `F` consists of those objects in the target category which are
+isomorphic to an object in the image of the function `F.obj`. In other words, this is the closure
+under isomorphism of the function `F.obj`.
+This is the "non-evil" way of describing the image of a functor.
+-/
+def ess_image (F : C ⥤ D) : set D := λ Y, ∃ (X : C), nonempty (F.obj X ≅ Y)
+
+/-- Get the witnessing object that `Y` is in the subcategory given by `F`. -/
+def ess_image.witness {Y : D} (h : Y ∈ F.ess_image) : C := h.some
+
+/-- Extract the isomorphism between `F.obj h.witness` and `Y` itself. -/
+def ess_image.get_iso {Y : D} (h : Y ∈ F.ess_image) : F.obj h.witness ≅ Y :=
+classical.choice h.some_spec
+
+/-- Being in the essential image is a "hygenic" property: it is preserved under isomorphism. -/
+lemma ess_image.of_iso {Y Y' : D} (h : Y ≅ Y') (hY : Y ∈ ess_image F) :
+  Y' ∈ ess_image F :=
+hY.imp (λ B, nonempty.map (≪≫ h))
+
+/--
+If `Y` is in the essential image of `F` then it is in the essential image of `F'` as long as
+`F ≅ F'`.
+-/
+lemma ess_image.of_nat_iso {F' : C ⥤ D} (h : F ≅ F') {Y : D} (hY : Y ∈ ess_image F) :
+  Y ∈ ess_image F' :=
+hY.imp (λ X, nonempty.map (λ t, h.symm.app X ≪≫ t))
+
+/-- Isomorphic functors have equal essential images. -/
+lemma ess_image_eq_of_nat_iso {F' : C ⥤ D} (h : F ≅ F') :
+  ess_image F = ess_image F' :=
+set.ext $ λ A, ⟨ess_image.of_nat_iso h, ess_image.of_nat_iso h.symm⟩
+
+/-- An object in the image is in the essential image. -/
+lemma obj_mem_ess_image (F : D ⥤ C) (Y : D) : F.obj Y ∈ ess_image F := ⟨Y, ⟨iso.refl _⟩⟩
+
+/-- The essential image of the functor `F` as an explicit subcategory of `D`. -/
+@[derive category]
+def essential_image (F : D ⥤ C) := {D // D ∈ F.ess_image}
```

Comment:
Should we have a simp lemma that simps the `coe_to_sort` of `ess_image` into `essential_image`?

Reply:
Maybe it shouldn't be simp... but at least the lemma seems good to have.

Reply:
Making this lemma I realised we could just define `essential_image` as the `coe_to_sort`, does this look okay?